File: 22.generate-parentheses.js
/*
 * @lc app=leetcode id=22 lang=javascript
 *
 * [22] Generate Parentheses
 */

// @lc code=start
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
  let res = [];
  // 这个比Grandyang 方法好多了
  const dfs = (path, left, right) => {
    if (left > n || left < right) return;
    if (left + right === 2 * n) {
      res.push(path);
    }
    dfs(path + "(", left + 1, right);
    dfs(path + ")", left, right + 1);
  };
  dfs("", 0, 0);
  return res;
};
// @lc code=end

File: 22.generate-parentheses.rb
#
# @lc app=leetcode id=22 lang=ruby
#
# [22] Generate Parentheses
#

# @lc code=start
# @param {Integer} n
# @return {String[]}
def generate_parenthesis(n)
  res = []
  backtrack = lambda { |path, left, right|
    return unless left >= right && left <= n
    return res << path.dup if left == n && right == n

    backtrack.call(path << '(', left + 1, right)
    path.chop!
    backtrack.call(path << ')', left, right + 1)
    path.chop!
  }
  backtrack.call('', 0, 0)
  res
end
# @lc code=end

generate_parenthesis(3)
